noir/noir-repo/compiler/noirc_evaluator/src/ssa.rs

```diff
@@ -70,31 +70,81 @@ pub(crate) fn optimize_into_acir(
 
     let last_array_uses = ssa.find_last_array_uses();
 
-    ssa.into_acir(brillig, abi_distinctness, &last_array_uses)
+    ssa.into_acir(&brillig, abi_distinctness, &last_array_uses)
 }
 
-/// Compiles the [`Program`] into [`ACIR`][acvm::acir::circuit::Circuit].
+/// Compiles the [`Program`] into [`ACIR``][acvm::acir::circuit::Program].
 ///
 /// The output ACIR is is backend-agnostic and so must go through a transformation pass before usage in proof generation.
 #[allow(clippy::type_complexity)]
 #[tracing::instrument(level = "trace", skip_all)]
-pub fn create_circuit(
+pub fn create_program(
     program: Program,
     enable_ssa_logging: bool,
     enable_brillig_logging: bool,
     force_brillig_output: bool,
-) -> Result<(Circuit, DebugInfo, Vec<Witness>, Vec<Witness>, Vec<SsaReport>), RuntimeError> {
+) -> Result<(AcirProgram, Vec<DebugInfo>, Vec<SsaReport>, Vec<Witness>, Vec<Witness>), RuntimeError>
+{
     let debug_variables = program.debug_variables.clone();
     let debug_types = program.debug_types.clone();
     let debug_functions = program.debug_functions.clone();
-    let func_sig = program.main_function_signature.clone();
+
+    let func_sigs = program.function_signatures.clone();
+
     let recursive = program.recursive;
-    let mut generated_acir = optimize_into_acir(
+    let generated_acirs = optimize_into_acir(
         program,
         enable_ssa_logging,
         enable_brillig_logging,
         force_brillig_output,
     )?;
+    assert_eq!(
+        generated_acirs.len(),
+        func_sigs.len(),
+        "The generated ACIRs should match the supplied function signatures"
+    );
+
+    let mut functions = vec![];
+    let mut debug_infos = vec![];
+    let mut warning_infos = vec![];
+    let mut main_input_witnesses = Vec::new();
+    let mut main_return_witnesses = Vec::new();
+    // For setting up the ABI we need separately specify main's input and return witnesses
+    let mut is_main = true;
+    for (acir, func_sig) in generated_acirs.into_iter().zip(func_sigs) {
+        let (circuit, debug_info, warnings, input_witnesses, return_witnesses) =
+            convert_generated_acir_into_circuit(
+                acir,
+                func_sig,
+                recursive,
+                // TODO: get rid of these clones
+                debug_variables.clone(),
+                debug_functions.clone(),
+                debug_types.clone(),
+            );
+        functions.push(circuit);
+        debug_infos.push(debug_info);
+        warning_infos.extend(warnings);
+        if is_main {
+            main_input_witnesses = input_witnesses;
+            main_return_witnesses = return_witnesses;
+        }
```

Comment:
`Circuit` exposes methods to get these witnesses right? We could query these after the fact.

Reply:
For `return_witness` those are just `return_values` in `Circuit` and as a BTreeSet is ordered this would be fine to use. However, for `input_witnesses` they do get split up into separate sets for `private_parameters` and `public_parameters`. ABI gen is expecting the input_witnesses in the order they were generated when computing the `param_witnesses` map in the `Abi`. It may be better to have a follow-up where we reconsider how ABI gen is done.